### PR TITLE
Blockfetch improvements needed before KeepAlive client can be enabled.

### LINKED ISSUE #2434: Limit the rate we make new blockfetch descisions
Limit blockfetch decision making to 100hz.

### PR DESCRIPTION
**Pull request by karknu**
Fixes #2434 #2433 and #2430 which are required for #1416.

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs [hunk 1]
```diff
@@ -91,10 +92,11 @@ data FetchMode =
   deriving (Eq, Show)
 
 
-type PeerInfo header extra =
+type PeerInfo header peer extra =
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs:95]: What about changing this into a record with strict fields.

Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs:95]: I'm not a fan of the "swizzel" functions used to access the different fields of the PeerInfo type, but I think that such a change is outside the scope of this PR.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs [hunk 2]
```diff
@@ -917,11 +943,20 @@ fetchRequestDecision FetchDecisionPolicy {
...
+  = Left $ FetchDeclineConcurrencyLimit
+             fetchMode maxConcurrentFetchPeers
+
+    -- If we're at the concurrency limit refuse any additional peers.
   | peerFetchReqsInFlight == 0
   , let maxConcurrentFetchPeers = case fetchMode of
                                     FetchModeBulkSync -> maxConcurrencyBulkSync
                                     FetchModeDeadline -> maxConcurrencyDeadline
-  , nConcurrentFetchPeers >= maxConcurrentFetchPeers
+  , nConcurrentFetchPeers == maxConcurrentFetchPeers
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs:959]: Could you elaborate why this change above?

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs [hunk 3]
```diff
@@ -854,11 +863,32 @@ fetchRequestDecisions fetchDecisionPolicy fetchMode chains =
...
                        peerFetchReqsInFlight)
       $ chains
 
 
+    -- Order the peers based on current PeerGSV. The top performing peers will be
+    -- permitted to go active even if we're above the desired maxConcurrentFetchPeers
+    -- which will cause us to switch smoothly from a slower to faster peers.
+    -- When switching from slow to faster peers we will be over the configured limit, but
+    -- PeerGSV is expected to be updated rather infrequently so the set of prefered peers should
+    -- be stable during 10s of second.
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs:876]: Looks good :)

File: ouroboros-network/test/Ouroboros/Network/BlockFetch/Examples.hs [hunk 4]
```diff
@@ -128,7 +128,7 @@ blockFetchExample1 decisionTracer clientStateTracer clientMsgTracer
           decisionTracer clientStateTracer
           (sampleBlockFetchPolicy1 blockHeap currentChainHeaders candidateChainHeaders)
           registry
-          (BlockFetchConfiguration 2 1 10)
+          (BlockFetchConfiguration 2 1 10 0.01)
```
Comment (coot) [inline ouroboros-network/test/Ouroboros/Network/BlockFetch/Examples.hs:131]: I would use the named fields to set up the fields, e.g.
```
BlockFetchConfiguration {
    bfcMaxConcurrencyBulkSync = 2,
    bfcMaxConcurrencyDeadline = 1,
    bfcMaxRequestsInflight = 10,
    bfcDecisionLoopInterval = 0.01
  }
```
This way it's easier to figure out the meaning of each value.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/State.hs [hunk 5]
```diff
@@ -67,19 +75,24 @@ fetchLogicIterations decisionTracer clientStateTracer
...
+      stateFingerprint' <- fetchLogicIteration
         decisionTracer clientStateTracer
         fetchDecisionPolicy
         fetchTriggerVariables
         fetchNonTriggerVariables
         stateFingerprint
+      end <- getMonotonicTime
+      let delta = diffTime end start
+      -- Limit descision making to once every decisionLoopInterval.
+      when (delta < decisionLoopInterval fetchDecisionPolicy) $ threadDelay delta
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/State.hs:94]: As you mentioned on our Slack thread: `threadDelay (decisionLoopInterval fetchDecisionPolicy - delta)`.

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs [hunk 6]
```diff
@@ -913,6 +913,9 @@ runThreadNetwork systemTime ThreadNetworkArgs
                   bfcMaxConcurrencyBulkSync = 1
                 , bfcMaxConcurrencyDeadline = 2
                 , bfcMaxRequestsInflight    = 10
+                , bfcDecisionLoopInterval   = 0.0 -- Mock testsuite can use sub-second slot
```
Comment (nfrisby) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs:916]: We decided to set it to `0` for now, which disables the rate limiter during the ThreadNet tests.

If this interval is greater than the slot length, then this will lead to test failures. In this case, some BF thread would eventually sleep for a duration that spans the next slot onset, but the tests currently assume that the nodes are more responsive than that -- roughly: "all nodes are fully synced by the end of every slot".

See https://input-output-rnd.slack.com/archives/CFKLUH4R0/p1595255785454400 for the discussion.
